File: iminling-core/src/main/java/com/iminling/core/filter/Filter.java
package com.iminling.core.filter;

import com.iminling.core.exception.BizException;
import org.springframework.core.Ordered;
import org.springframework.core.annotation.AnnotationUtils;
import org.springframework.web.method.HandlerMethod;

import javax.servlet.http.HttpServletRequest;
import java.lang.annotation.Annotation;

public interface Filter extends Ordered {

    void doFilter(HandlerMethod handlerMethod, HttpServletRequest request) throws BizException;

    default <T extends Annotation> T getAnnotation(HandlerMethod handlerMethod, Class<T> clazz) {
        T annotation = AnnotationUtils.findAnnotation(handlerMethod.getMethod(), clazz);
        if (annotation == null) {
            annotation = handlerMethod.getBeanType().getAnnotation(clazz);
        }
        return annotation;
    }
}

File: iminling-core/src/main/kotlin/com/iminling/core/config/filter/Filter.kt
package com.iminling.core.config.filter

import com.iminling.core.exception.BizException
import org.springframework.core.Ordered
import org.springframework.core.annotation.AnnotationUtils
import org.springframework.web.method.HandlerMethod
import javax.servlet.http.HttpServletRequest

/**
 * @author  yslao@outlook.com
 * @since  2021/11/28
 */
// @JvmDefaultWithoutCompatibility
interface Filter: Ordered {

    @Throws(BizException::class)
    fun doFilter(handlerMethod: HandlerMethod, request: HttpServletRequest)

    fun <T : Annotation> getAnnotation(handlerMethod: HandlerMethod, clazz: Class<T>): T? {
        return AnnotationUtils.findAnnotation(handlerMethod.method, clazz)
            ?: handlerMethod.beanType.getAnnotation(clazz)
    }

}